src/OpenLoco/src/Vehicles/VehicleHead.cpp

```diff
@@ -3401,15 +3424,239 @@ namespace OpenLoco::Vehicles
+                    {
+                        continue;
+                    }
+                    if (elStation->isGhost() || elStation->isAiAllocated())
+                    {
+                        continue;
+                    }
+                    if (elStation->baseZ() != pos.z / World::kSmallZStep)
+                    {
+                        continue;
+                    }
+                    return elStation;
+                }
+                return nullptr;
+            }();
+            if (elStation == nullptr)
+            {
+                continue;
+            }
+            if (elStation->stationType() != StationType::docks)
+            {
+                continue;
+            }
+            if (elStation->isFlag6())
+            {
+                continue;
+            }
+
+            const auto* dockObj = ObjectManager::get<DockObject>(elStation->objectId());
+            const auto boatPos = Math::Vector::rotate(dockObj->boatPosition, elStation->rotation()) + pos;
+            auto dockTarget = WaterPathingResult(
+                boatPos + World::Pos2{ 32, 32 },
+                stationId,
+                World::Pos3{ pos.x, pos.y, Numerics::floor2(pos.z, 4) });
+            return dockTarget;
+        }
+        return std::nullopt;
+    }
+
+    struct PathFindingResult
+    {
+        uint16_t bestScore;
+        uint8_t cost;
+
+        constexpr auto operator<=>(const PathFindingResult& rhs) const = default;
+    };
+
+    // 0x00428237
+    static PathFindingResult waterPathfindToTarget(const World::TilePos2 tilePos, const MicroZ waterMicroZ, const World::TilePos2 targetOrderPos, const NearbyBoats& nearbyVehicles, uint8_t cost, const PathFindingResult& bestResult)
+    {
+        PathFindingResult result = bestResult;
+        if (!validCoords(tilePos))
+        {
+            return result;
+        }
+        auto tile = TileManager::get(tilePos);
+        auto* elSurface = tile.surface();
+        if (elSurface->water() != waterMicroZ)
+        {
+            return result;
+        }
+        if (!elSurface->isLast())
+        {
+            auto* elObsticle = elSurface->next();
+            if (elObsticle != nullptr && !elObsticle->isGhost() && !elObsticle->isAiAllocated())
+            {
+                if (elObsticle->baseZ() / kMicroToSmallZStep - waterMicroZ < 1)
+                {
+                    return result;
+                }
+            }
+        }
+
+        const auto nearbyIndex = tilePos - nearbyVehicles.startTile;
+        // Vanilla made a mistake here so we only check nearby tiles if in range
+        // TODO: When we diverge just change the cost check to >= 6 or increase the search result to 18x18
+        if (nearbyIndex.x >= 0 && nearbyIndex.x < 16 && nearbyIndex.y >= 0 && nearbyIndex.y < 16)
+        {
+            // assert(nearbyIndex.x >= 0 && nearbyIndex.x < 16);
+            // assert(nearbyIndex.y >= 0 && nearbyIndex.y < 16);
```

Comment:
It looks like this is covered by the `if` expression. Can these be removed?

Reply:
They could be removed now. They were sort of linked to the TODO. In vanilla it just walked passed the end of the array.